cfgrammar/src/lib/header.rs

```diff
@@ -0,0 +1,186 @@
+                                Ok((
+                                    ValBind::Bind(
+                                        (key_name, key_span),
+                                        Val::ArgLike(path_val, arg),
+                                    ),
+                                    i,
+                                ))
+                            } else {
+                                Err(HeaderError::ExpectedToken(')', Span::new(i, i)))
+                            }
+                        } else {
+                            Ok((
+                                ValBind::Bind((key_name, key_span), Val::PathLike(path_val)),
+                                i,
+                            ))
+                        }
+                    }
+                }
+            } else {
+                Ok((ValBind::TrueKey(key_name, key_span), i))
+            }
+        }
+    }
+
+    fn parse_path(&'_ self, mut i: usize) -> Result<(Path<'_>, usize), HeaderError> {
+        let (name, j) = self.parse_name(i)?;
+        let name_span = Span::new(i, j);
+        i = self.parse_ws(j);
+        if let Some(j) = self.lookahead_is("::", i) {
+            i = self.parse_ws(j);
+            let (scoped_val, j) = self.parse_name(i)?;
+            let scoped_span = Span::new(i, j);
+            i = self.parse_ws(j);
+            Ok((
+                Path::Scoped((name, name_span), (scoped_val, scoped_span)),
+                i,
+            ))
+        } else {
+            Ok((Path::Ident(name, name_span), i))
+        }
+    }
+
+    pub fn new(src: &'input str, required: bool) -> Self {
+        Self { src, required }
+    }
+
+    pub fn parse(&'_ self) -> Result<(Vec<ValBind<'_>>, usize), HeaderError> {
+        if let Some(mut i) = self.lookahead_is(MAGIC, 0) {
+            let mut vals = vec![];
+            i = self.parse_ws(i);
+            if let Some(j) = self.lookahead_is("{", i) {
+                i = self.parse_ws(j);
+                while self.lookahead_is("}", i).is_none() {
+                    let (val, j) = self.parse_valbind(i)?;
+                    vals.push(val);
+                    if let Some(j) = self.lookahead_is(",", j) {
+                        i = self.parse_ws(j);
+                        continue;
+                    } else {
+                        return Ok((vals, j));
+                    }
+                }
+                Ok((vals, i))
+            } else {
+                Err(HeaderError::ExpectedToken('{', Span::new(i, i)))
+            }
+        } else if self.required {
+            Err(HeaderError::MissingGrmtoolsSection(Span::new(0, 0)))
+        } else {
+            Ok((vec![], 0))
+        }
+    }
+
+    fn parse_name(&self, i: usize) -> Result<(&str, usize), HeaderError> {
+        match RE_NAME.find(&self.src[i..]) {
+            Some(m) => {
+                assert_eq!(m.start(), 0);
+                Ok((&self.src[i..i + m.end()], i + m.end()))
```

Comment:
Should consider doing
```suggestion
                Ok((&self.src[i..i + m.end()].to_lowercase(), i + m.end()))
```

Reply:
Don't think that is possible, I had forgot that this is borrowed, so to_lowercase() would return a temporary.
I think callers will have to do it, unless we want to change these to `Strings` so we can lowercase them.